tests/CMakeLists.txt

```diff
@@ -79,4 +79,35 @@ add_executable(knnepsilon knnepsilon.cpp)
 target_link_libraries(knnepsilon ${LIB_NAME} ${EXTRA_LIBS} ${Boost_LIBRARIES})
 
 add_executable(knnbucketsize knnbucketsize.cpp)
-target_link_libraries(knnbucketsize ${LIB_NAME} ${EXTRA_LIBS} ${Boost_LIBRARIES})
\ No newline at end of file
+target_link_libraries(knnbucketsize ${LIB_NAME} ${EXTRA_LIBS} ${Boost_LIBRARIES})
+
+# Ensure that users cannot instantiate a tree with wrong matrix types.
+macro(try_compile_disallowed_types MAT_TYPE TREE_TYPE EXPECT)
+  try_compile(COMPILE_SUCCEEDED
+                "${CMAKE_CURRENT_BINARY_DIR}"
+                "${CMAKE_CURRENT_SOURCE_DIR}/invalid_matrix_types_disallowed.cpp"
+                COMPILE_DEFINITIONS "-D${MAT_TYPE} -D${TREE_TYPE} -I${CMAKE_CURRENT_SOURCE_DIR}/../ -I${EIGEN_INCLUDE_DIR} -l${LIB_NAME}"
+                CMAKE_FLAGS "-DCMAKE_CXX_LINK_EXECUTABLE='echo not linking now...'"
+                OUTPUT_VARIABLE OUTPUT)
+
+  if (${EXPECT} STREQUAL "TRUE") 
+    if(NOT COMPILE_SUCCEEDED)
+      message( FATAL_ERROR "Compiling a test executable failed." )
```

Comment:
Will this be detected by Jenkins as an error? I would assume so, but I'm not sure.